qiskit/transpiler/target.py

```diff
@@ -0,0 +1,387 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+A target object represents the minimum set of information the transpiler needs
+from a backend
+"""
+
+import logging
+
+import retworkx as rx
+
+from qiskit.transpiler.coupling import CouplingMap
+from qiskit.transpiler.exceptions import CouplingError
+from qiskit.transpiler.instruction_durations import InstructionDurations
+
+logger = logging.getLogger(__name__)
+
+
+class InstructionProperties:
+    """A representation of the properties of a gate implementation."""
+
+    __slots__ = ("length", "error", "properties")
+
+    def __init__(self, length=None, error=None, properties=None):
+        self.length = length
+        self.error = error
+        self.properties = properties
+
+
+class Target:
+    """
+    A description of gates on a backend. It exists around a central mapping of
+    :class:`~qiskit.circuit.Instruction` objects to their properties on the
+    device. As a basic example, lets assume your device is two qubits, supports
+    :class:`~qiskit.circuit.library.UGate` on both qubits and
+    :class:`~qiskit.circuit.library.CXGate` in both directions you would create
+    the gate map like::
+
+        from qiskit.transpiler import Target
+        from qiskit.circuit.library import UGate, CXGate
+        from qiskit.circuit import Parameter
+
+        gmap = Target()
+        theta = Parameter('theta')
+        phi = Parameter('phi')
+        lam = Parameter('lambda')
+        u_props = {
+            (0,): InstructionProperties(length=5.23e-8, error=0.00038115),
+            (1,): InstructionProperties(length=4.52e-8, error=0.00032115),
+        }
+        gmap.add_gate(UGate(theta, phi, lam), [(0,), (1,)], properties=u_props)
```

Comment:
The behavior of this object is pretty close to what `InstructionScheduleMap` implements for pulse (If you add a schedule to `InstructionProperties`, this will be the upper compatible of instmap. Do we integrate instmap into this object and gradually deprecate the instmap, rather than having two similar containers?

```suggestion
        gmap.add_instruction(UGate(theta, phi, lam), [(0,), (1,)], properties=u_props)
```

Reply:
yeah I like that idea I'll add a pulse field to the `InstructionProperties` object

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/5885/commits/1dc54d739df7d3aedcb83676fc47e80ad47cb16e